app/javascript/components/server-components/ProductEditPage.tsx

```diff
@@ -195,8 +202,22 @@ const ProductEditPage = (props: Props) => {
       saving,
       contentUpdates,
       setContentUpdates,
+      availableDiscountCodes,
+      setAvailableDiscountCodes,
     }),
-    [product, updateProduct, existingFiles, setExistingFiles],
+    [
+      product,
+      updateProduct,
+      existingFiles,
+      setExistingFiles,
+      availableDiscountCodes,
+      setAvailableDiscountCodes,
+      currencyType,
+      setCurrencyType,
+      saving,
+      contentUpdates,
+      setContentUpdates,
+    ],
```

Comment:
ProductEdit context now carries availableDiscountCodes + setter. We only hydrate the currently selected code from Rails; everything else comes via the new endpoint to avoid shipping thousands of codes on first load